test/Main.hs

```diff
@@ -1,74 +1,191 @@
+                                            ]
+                                        )
+                                    ,
+                                        ( "SERVER"
+                                        , [("port", "6667"), ("hostname", "localhost")]
+                                        )
+                                    ]
+                            , iniGlobals = []
+                            }
+                        )
+
+            it "parses file with globals" $
+                parseIni
+                    "# Some comment.\n\
+                    \port=6667\n\
+                    \hostname=localhost\n\
+                    \[AUTH]\n\
+                    \user=hello\n\
+                    \pass=world\n\
+                    \# Salt can be an empty string.\n\
+                    \salt="
+                    `shouldBe` Right
+                        ( Ini
+                            { iniSections =
+                                HM.fromList
+                                    [
+                                        ( "AUTH"
+                                        ,
+                                            [ ("user", "hello")
+                                            , ("pass", "world")
+                                            , ("salt", "")
+                                            ]
+                                        )
+                                    ]
+                            , iniGlobals =
+                                [("port", "6667"), ("hostname", "localhost")]
+                            }
+                        )
+
+            it "fails to parse file with invalid keys" $
+                parseIni
+                    "Name=Foo\n\
+                    \Name[en_GB]=Fubar"
+                    `shouldBe` Left "Failed reading: Name[en_GB]=Fubar"
+
+            it "parses file ending with comments" $
+                parseIni
+                    "[default]\n\
+                    \a = 1\n\
+                    \\n\
+                    \#[staging-PI]\n\
+                    \#a = 2\n"
+                    `shouldBe` Right
+                        ( Ini
+                            { iniSections = HM.fromList [("default", [("a", "1")])]
+                            , iniGlobals = []
+                            }
+                        )
+
+            it "parses file with globals only" $
+                parseIni
+                    "global1 = hello\n\
+                    \global2 = 123\n\
+                    \# An end of file comment here"
+                    `shouldBe` Right
+                        ( Ini
+                            { iniSections = HM.empty
+                            , iniGlobals = [("global1", "hello"), ("global2", "123")]
+                            }
+                        )
+
+            it "parses empty file" $
+                parseIni "" `shouldBe` Right mempty
+
+            it "roundtrips with printIniWith" $
```

Comment:
Added a roundtrip property test, as my initial attempt at fix didn't catch another broken case (globals only + comments at the end)